### PR TITLE
Introduce i64 priority for better compatibility

### PR DESCRIPTION
**Pull request by ethercflow**
ref tikv/tikv#12876

Address https://github.com/tikv/raft-rs/pull/493#issuecomment-1306861338

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

### CODE REVIEW

File: proto/proto/eraftpb.proto [hunk 1]
```diff
@@ -89,6 +89,12 @@ message Message {
     uint64 reject_hint = 11;
     bytes context = 12;
     uint64 priority = 14;
+    // If this new field is not set, then use the above old field; otherwise
+    // use the new field. When broadcasting request vote, both fields are
+    // set if the priority is larger than 0. This change is not a fully
+    // compatible change, but it makes minimal impact that only new priority
+    // is not recognized by the old nodes during rolling update.
+    int64 new_priority = 16;
```
Comment (Connor1996) [inline proto/proto/eraftpb.proto:97]: maybe call it `priority_v2`

**Review by Connor1996 — COMMENTED** [review]
rest LGTM

**Review by Connor1996 — APPROVED** [review]
LGTM. And please give a better PR title.

File: src/raft.rs [hunk 2]
```diff
@@ -1457,7 +1460,13 @@ impl<T: Storage> Raft<T> {
                 // ...and we believe the candidate is up to date.
                 if can_vote
                     && self.raft_log.is_up_to_date(m.index, m.log_term)
-                    && (m.index > self.raft_log.last_index() || self.priority <= m.priority)
+                    && (m.index > self.raft_log.last_index()
+                        || self.priority
+                            <= if m.priority != 0 {
```
Comment (BusyJay) [inline src/raft.rs:1465]: Better extract as a function.

File: src/raft.rs [hunk 3]
```diff
@@ -1457,7 +1460,13 @@ impl<T: Storage> Raft<T> {
...
+                                m.priority
+                            } else {
+                                m.deprecated_priority as i64
```
Comment (BusyJay) [inline src/raft.rs:1468]: And it should saturated.

**Review by Connor1996 — APPROVED** [review]
LGTM

File: src/raft.rs [hunk 4]
```diff
@@ -303,6 +303,14 @@ fn new_message(to: u64, field_type: MessageType, from: Option<u64>) -> Message {
     m
 }
 
+fn get_priority(m: &Message) -> i64 {
+    if m.priority != 0 {
+        m.priority
+    } else {
+        0_i64.saturating_add_unsigned(m.deprecated_priority)
```
Comment (BusyJay) [inline src/raft.rs:310]: This is a nightly only API. You can do that within stable `try_from.unwrap_or(max)`

Comment (ethercflow) [reply src/raft.rs]: Fixed, thanks!

**Review by BusyJay — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]

**Comment by ethercflow** [comment]
@BusyJay Lint has been fixed, please help merge when you have time, thanks!

**Pull request merged**